File: src/windows/Maple.WinUI/MapleUI/MoleculesL2/MuiResponseViewer.cs
using System.Collections.Generic;
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Controls;
using Maple.UI.Atoms;

namespace Maple.UI
{
    /// <summary>
    /// Maple.UI Response Viewer molecule (unified-component-catalog.md §3,
    /// "Response Viewer" row: "Formatted response with status", built from
    /// Code Block, Badge, Tabs) — an HTTP status badge, a Body/Headers
    /// <see cref="MuiTabs"/> switch, and a <see cref="MuiCodeBlock"/>
    /// showing whichever tab is active. Ports
    /// `mui-response-viewer.component.ts`'s <c>statusVariant</c> exactly
    /// (status &lt; 400 → Signal, else Count — the web port's own choice,
    /// not restyled here).
    /// </summary>
    public sealed class MuiResponseViewer : ContentControl
    {
        private static readonly IReadOnlyList<MuiTab> Tabs = new[]
        {
            new MuiTab("body", "Body"),
            new MuiTab("headers", "Headers"),
        };

        public static readonly DependencyProperty StatusProperty =
            DependencyProperty.Register(nameof(Status), typeof(int), typeof(MuiResponseViewer),
                new PropertyMetadata(200, (d, _) => ((MuiResponseViewer)d).Rebuild()));

        public static readonly DependencyProperty StatusTextProperty =
            DependencyProperty.Register(nameof(StatusText), typeof(string), typeof(MuiResponseViewer),
                new PropertyMetadata(string.Empty, (d, _) => ((MuiResponseViewer)d).Rebuild()));

        public static readonly DependencyProperty BodyProperty =
            DependencyProperty.Register(nameof(Body), typeof(string), typeof(MuiResponseViewer),
                new PropertyMetadata(string.Empty, (d, _) => ((MuiResponseViewer)d).Rebuild()));

        public static readonly DependencyProperty HeadersProperty =
            DependencyProperty.Register(nameof(Headers), typeof(string), typeof(MuiResponseViewer),
                new PropertyMetadata(string.Empty, (d, _) => ((MuiResponseViewer)d).Rebuild()));

        public static readonly DependencyProperty ActiveIdProperty =
            DependencyProperty.Register(nameof(ActiveId), typeof(string), typeof(MuiResponseViewer),
                new PropertyMetadata("body", (d, e) => { ((MuiResponseViewer)d)._tabs.ActiveId = (string)e.NewValue; ((MuiResponseViewer)d).Rebuild(); }));

        public int Status
        {
            get => (int)GetValue(StatusProperty);
            set => SetValue(StatusProperty, value);
        }

        public string StatusText
        {
            get => (string)GetValue(StatusTextProperty);
            set => SetValue(StatusTextProperty, value);
        }

        public string Body
        {
            get => (string)GetValue(BodyProperty);
            set => SetValue(BodyProperty, value);
        }

        public string Headers
        {
            get => (string)GetValue(HeadersProperty);
            set => SetValue(HeadersProperty, value);
        }

        public string ActiveId
        {
            get => (string)GetValue(ActiveIdProperty);
            set => SetValue(ActiveIdProperty, value);
        }

        private readonly StackPanel _root = new() { Orientation = Orientation.Vertical, Spacing = 10 };
        private readonly StackPanel _header = new() { Orientation = Orientation.Horizontal, Spacing = 12 };
        private readonly MuiBadge _statusBadge = new();
        private readonly MuiTabs _tabs = new() { Tabs = Tabs, ActiveId = "body" };
        private readonly MuiCodeBlock _codeBlock = new();

        public MuiResponseViewer()
        {
            _header.Children.Add(_statusBadge);
            _header.Children.Add(_tabs);
            _root.Children.Add(_header);
            _root.Children.Add(_codeBlock);
            Content = _root;
            IsTabStop = false;

            _tabs.SelectionChanged += (_, id) => ActiveId = id;

            Rebuild();
        }

        private void Rebuild()
        {
            _statusBadge.Variant = Status < 400 ? MuiBadgeVariant.Signal : MuiBadgeVariant.Count;
            _statusBadge.Value = $"{Status} {StatusText}".Trim();

            _codeBlock.Code = ActiveId == "headers" ? Headers : Body;
        }
    }
}

File: src/apple/Packages/MapleUI/Sources/MapleUI/MoleculesL2/MuiResponseViewer.swift
// MuiResponseViewer.swift — Maple UI Molecules-L2 (unified-component-
// catalog.md §3). Formatted response with status, built from Code Block,
// Badge, Tabs.

import SwiftUI

public struct MuiResponseViewer: View {
    public let status: Int
    public let statusText: String
    public let body_: String
    public let headers: String
    @Binding public var activeId: String

    private static let tabs: [MuiTab] = [
        MuiTab(id: "body", label: "Body"),
        MuiTab(id: "headers", label: "Headers"),
    ]

    public init(status: Int, statusText: String = "", body: String, headers: String = "", activeId: Binding<String> = .constant("body")) {
        self.status = status
        self.statusText = statusText
        self.body_ = body
        self.headers = headers
        self._activeId = activeId
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: MuiTokens.spacingSm) {
            HStack {
                MuiBadge(variant: Self.statusVariant(status: status), value: Self.statusLabel(status: status, statusText: statusText))
                Spacer()
                MuiTabs(tabs: Self.tabs, activeId: $activeId, accessibilityLabel: "Response tabs")
            }
            MuiCodeBlock(code: Self.activeContent(activeId: activeId, body: body_, headers: headers))
        }
    }

    /// The badge caption combining status code and reason phrase. Public +
    /// static so this is unit-testable without rendering a view.
    public static func statusLabel(status: Int, statusText: String) -> String {
        "\(status) \(statusText)".trimmingCharacters(in: .whitespaces)
    }

    /// `.signal` for a successful (< 400) response, `.count` otherwise.
    /// Public + static so this is unit-testable without rendering a view.
    public static func statusVariant(status: Int) -> MuiBadgeVariant {
        status < 400 ? .signal : .count
    }

    /// The code block's content for the active tab. Public + static so
    /// this is unit-testable without rendering a view.
    public static func activeContent(activeId: String, body: String, headers: String) -> String {
        activeId == "headers" ? headers : body
    }
}

#Preview("MuiResponseViewer") {
    struct Demo: View {
        @State private var activeId = "body"

        var body: some View {
            MuiResponseViewer(
                status: 200,
                statusText: "OK",
                body: "{\n  \"id\": \"IMG_0042\"\n}",
                headers: "content-type: application/json",
                activeId: $activeId
            )
            .padding()
            .frame(width: 340)
            .background(MuiTokens.bg)
        }
    }
    return Demo()
}
